ipld/read.go

```diff
@@ -18,190 +16,131 @@ import (
+			if err != nil {
+				return err
+			}
 
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+			leaves, err := GetLeaves(ctx, dag, subtreeRootHash, uint32(len(roots)))
+			if err != nil {
+				return err
+			}
+			for leafIdx, leaf := range leaves {
+				shareData := leaf.RawData()[1:]
+				// it's not needed to store data for cols
+				// as we are fetching data from the same share for rows and cols
+				if isRow {
+					dataSquare[(i*len(roots)*2)+leafIdx] = shareData[NamespaceSize:]
+				}
+			}
+			return err
+		})
+	}
 
-	return &shareCounter{
-		shares:          make(map[index][]byte),
-		edsWidth:        edsWidth,
-		minSharesNeeded: minSharesNeeded,
-		shareChan:       make(chan indexedShare, 512),
-		errc:            make(chan error, 1),
-		ctx:             ctx,
-		cancel:          cancel,
+	for i := 0; i < len(roots); i++ {
+		fetcher(i)
 	}
+	return errGroup.Wait()
 }
 
-// retrieveLeaf uses GetLeafData to fetch a single leaf and counts that leaf
-func (sc *shareCounter) retrieveShare(
-	rootCid cid.Cid,
-	isRow bool,
-	axisIdx uint32,
-	idx uint32,
-	dag ipld.NodeGetter,
-) {
-	data, err := GetLeafData(sc.ctx, rootCid, idx, sc.edsWidth, dag)
+// GetSubtreeLeaves returns only one subtree - left or right
+func GetSubtreeLeaves(ctx context.Context, root cid.Cid, dag ipld.NodeGetter, isLeftSubtree bool) (cid.Cid, error) {
+	nd, err := dag.Get(ctx, root)
 	if err != nil {
-		select {
-		case <-sc.ctx.Done():
-		case sc.errc <- err:
-		}
+		return cid.Cid{}, err
 	}
-
-	if len(data) < plugin.ShareSize {
-		return
+	links := nd.Links()
+	if isLeftSubtree || len(links) == 1 {
+		return links[0].Cid, nil
 	}
 
-	// switch the row and col indexes if needed
-	rowIdx := idx
-	colIdx := axisIdx
-	if isRow {
-		rowIdx = axisIdx
-		colIdx = idx
-	}
+	return links[1].Cid, nil
+}
 
-	select {
-	case <-sc.ctx.Done():
-	case sc.shareChan <- indexedShare{data: data[NamespaceSize:], index: index{row: rowIdx, col: colIdx}}:
+func GetLeaves(ctx context.Context, dag ipld.NodeGetter, root cid.Cid, size uint32) ([]ipld.Node, error) {
+	leaves, err := getLeaves(ctx, dag, root, make([]ipld.Node, 0, size))
+	if err != nil {
+		return nil, err
 	}
-}
 
-// wait until enough data has been collected, the timeout has been reached, or
-// too many errors are encountered
-func (sc *shareCounter) wait() error {
-	defer sc.cancel()
+	return leaves, nil
+}
 
-	for {
-		select {
-		case <-sc.ctx.Done():
-			err := sc.ctx.Err()
-			if err == context.DeadlineExceeded {
-				return ErrRetrieveTimeout
-			}
-			return err
-		case share := <-sc.shareChan:
-			_, has := sc.shares[share.index]
-			// add iff it does not already exists
-			if !has {
-				sc.shares[share.index] = share.data
-				sc.counter++
-				// check finishing condition
-				if sc.counter >= sc.minSharesNeeded {
-					return nil
-				}
-			}
+// getLeaves recursively starts going down to find all leafs from the given root
```

Comment:
What is the difference between `GetLeaves` and `getLeaves`? Looks like the latter is the implementation? They should probably have different names.

Reply:
`GetLeaves` is a wrapper over `getLeaves` but I think they both have correct namings. I've decided to move more complex logic to a separate function.